src/backend/impl/src/services/proposal_service.rs

```diff
@@ -0,0 +1,284 @@
+            })?;
+
+        match (proposal.state, state.clone()) {
+            (ReviewPeriodState::InProgress, ReviewPeriodState::Completed) => {
+                proposal.state = state;
+
+                self.proposal_repository.update_proposal(&id, proposal)
+            }
+            _ => Err(ApiError::invalid_argument(
+                "Invalid proposal state transition",
+            )),
+        }
+    }
+
+    async fn fetch_and_save_nns_proposals(&self) -> Result<(), ApiError> {
+        let nns_governance_canister =
+            GovernanceCanisterService(Principal::from_str(NNS_GOVERNANCE_CANISTER_ID).unwrap());
+
+        let ListProposalInfoResponse {
+            proposal_info: proposals,
+        } = match nns_governance_canister
+            .list_proposals(ListProposalInfo {
+                include_reward_status: vec![],
+                omit_large_fields: Some(true),
+                before_proposal: None,
+                limit: 20,
+                exclude_topic: vec![
+                    Topic::Unspecified.into(),
+                    Topic::NeuronManagement.into(),
+                    Topic::ExchangeRate.into(),
+                    Topic::NetworkEconomics.into(),
+                    Topic::Governance.into(),
+                    Topic::NodeAdmin.into(),
+                    Topic::ParticipantManagement.into(),
+                    Topic::SubnetManagement.into(),
+                    Topic::Kyc.into(),
+                    Topic::NodeProviderRewards.into(),
+                    Topic::SnsDecentralizationSale.into(),
+                    Topic::SubnetReplicaVersionManagement.into(),
+                    Topic::SnsAndCommunityFund.into(),
+                    Topic::ApiBoundaryNodeManagement.into(),
+                ],
+                include_all_manage_neuron_proposals: Some(false),
+                include_status: vec![ProposalStatus::Open.into()],
+            })
+            .await
+        {
+            Ok(response) => response.0,
+            Err(err) => {
+                return Err(ApiError::internal(&format!(
+                    "Failed to fetch proposals: {:?}",
+                    err
+                )))
+            }
+        };
+
+        for nns_proposal in proposals {
+            let proposal = match Proposal::try_from(nns_proposal) {
+                Ok(proposal) => proposal,
+                Err(err) => {
+                    return Err(ApiError::internal(&format!(
+                        "Failed to map NNS proposal: {:?}",
+                        err
+                    )))
+                }
+            };
+
+            match self
+                .proposal_repository
+                .get_proposals()
+                .iter()
+                .find(|(_, p)| p.nervous_system.id() == proposal.nervous_system.id())
+            {
+                Some(_) => {
+                    // TODO: decide what to do if the proposal already exists
```

Comment:
I think we don't need to do anything if it already exists, so we could just do this:
```
if self.proposal_repository
        .get_proposals()
        .iter()
        .find(|(_, p)| p.nervous_system.id() == proposal.nervous_system.id())
        .is_none() {
    self.proposal_repository.create_proposal(proposal).await?;
}
```